pkg/deploy/stack_deployer.go

```diff
@@ -54,35 +60,140 @@ type latticeServiceStackDeployer struct {
 		k8sClient:             k8sClient,
 		latticeServiceManager: lattice.NewServiceManager(log, cloud),
-		targetGroupManager:    lattice.NewTargetGroupManager(log, cloud),
+		targetGroupManager:    tgMgr,
 		targetsManager:        lattice.NewTargetsManager(log, cloud),
 		listenerManager:       lattice.NewListenerManager(log, cloud),
 		ruleManager:           lattice.NewRuleManager(log, cloud),
 		dnsEndpointManager:    externaldns.NewDnsEndpointManager(log, k8sClient),
-		svcExportTgBuilder:    gateway.NewSvcExportTargetGroupBuilder(log, k8sClient),
-		svcBuilder:            gateway.NewLatticeServiceBuilder(log, k8sClient, brTgBuilder),
+		svcExportTgBuilder:    tgSvcExpBuilder,
+		svcBuilder:            svcBuilder,
 	}
 }
 
+type TgGcCycleFn = func(context.Context) (TgGcResult, error)
+
+func NewTgGcFn(tgSynth *lattice.TargetGroupSynthesizer) TgGcCycleFn {
+	return func(ctx context.Context) (TgGcResult, error) {
+		t0 := time.Now()
+		results, err := tgSynth.SynthesizeUnusedDelete(ctx)
+		if err != nil {
+			return TgGcResult{}, err
+		}
+		succ := 0
+		for _, res := range results {
+			if res.Err == nil {
+				succ += 1
+			}
+		}
+		return TgGcResult{
+			att:      len(results),
+			succ:     succ,
+			duration: time.Since(t0),
+		}, nil
+	}
+}
+
+type TgGc struct {
+	lock    sync.Mutex
+	log     gwlog.Logger
+	ctx     context.Context
+	isDone  atomic.Bool
+	ivl     time.Duration
+	cycleFn TgGcCycleFn
+}
+
+type TgGcResult struct {
+	// number deletion attempts
+	att int
+	// number of successful deletions
+	succ int
+	// cycle duration
+	duration time.Duration
+}
+
+func (gc *TgGc) start() {
+	ticker := time.NewTicker(gc.ivl)
+	go func() {
+		for {
+			select {
+			case <-gc.ctx.Done():
+				gc.log.Info("stop GC, ctx is done")
+				gc.isDone.Store(true)
+				return
+			case <-ticker.C:
+				gc.cycle()
+			}
+		}
+	}()
+}
+
+func (gc *TgGc) cycle() {
+	defer func() {
+		if r := recover(); r != nil {
+			gc.log.Errorf("gc cycle panic: %s", r)
+		}
+		gc.lock.Unlock()
+	}()
+	gc.lock.Lock()
+	res, err := gc.cycleFn(gc.ctx)
+	if err != nil {
+		gc.log.Debugf("gc cycle error: %s", err)
+	}
+	gc.log.Debugw("tg gc result",
+		"delete_attempts", res.att,
+		"delete_success", res.succ,
+		"duration", res.duration,
+	)
+}
+
 func (d *latticeServiceStackDeployer) Deploy(ctx context.Context, stack core.Stack) error {
 	targetGroupSynthesizer := lattice.NewTargetGroupSynthesizer(d.log, d.cloud, d.k8sClient, d.targetGroupManager, d.svcExportTgBuilder, d.svcBuilder, stack)
 	targetsSynthesizer := lattice.NewTargetsSynthesizer(d.log, d.cloud, d.targetsManager, stack)
 	serviceSynthesizer := lattice.NewServiceSynthesizer(d.log, d.latticeServiceManager, d.dnsEndpointManager, stack)
 	listenerSynthesizer := lattice.NewListenerSynthesizer(d.log, d.listenerManager, stack)
 	ruleSynthesizer := lattice.NewRuleSynthesizer(d.log, d.ruleManager, d.targetGroupManager, stack)
 
+	defer func() {
+		tgGc.lock.Unlock()
+	}()
+	tgGc.lock.Lock()
```

Comment:
Can you add some comments here for the reason we need to do Lock and Unlock here?

Reply:
ok